File: packages/xyd-opensdk-node/src/resource.ts
import type { Method, NamedType, Param, ResolvedSdkBehavior, Resource, TypeRef } from '@xyd-js/opensdk-core';
import { type OperationPlan, type PageName, planOperation } from '@xyd-js/opensdk-framework';

import { jsDoc, propKey, unionDecodeName } from './model';
import { camelCase, nodeMethodName, pascalCase, singularPascalCase, slug } from './naming';
import { ModelRefs, bodyFields, isBinaryTypeRef, nodeType } from './nodetype';

/** Per-file render context: the IR symbol table + resolved runtime behavior. */
export interface NodeCtx {
  types: Map<string, NamedType>;
  behavior: ResolvedSdkBehavior;
}

/** Per-file import trackers (model types, decode fns, page containers, joinCsv). */
interface FileUses {
  refs: ModelRefs;
  decodeNames: Set<string>;
  pages: Set<PageName>;
  needsJoinCsv: boolean;
  hasNested: boolean;
}

/**
 * The globally-unique resource CLASS name, path-qualified by the FULL segment
 * chain (openai-node keeps the plural: `Pets`, `Batches`). Two `photos`
 * resources under different parents get distinct names (`PetsPhotos` vs
 * `UsersPhotos`) so one never shadows the other in the single generated file.
 */
export function resourceClassName(segments: string[]): string {
  return segments.map(pascalCase).join('');
}

/** The params-type qualifier (SINGULARIZED, openai-node's `PetCreateParams` shape). */
function paramsQualifier(segments: string[]): string {
  return segments.map(singularPascalCase).join('');
}

/** Emit `src/resources/index.ts`: the barrel re-exporting every top-level resource file. */
export function renderResourcesIndexFile(resources: Resource[]): string {
  const lines = resources.map((r) => `export * from './${slug(r.name) || 'resource'}';`);
  return `${lines.join('\n')}\n`;
}

/** Emit one top-level resource (and its whole subtree) into a single file. */
export function renderResourceFile(resource: Resource, ctx: NodeCtx): { path: string; content: string } {
  const uses: FileUses = {
    refs: new ModelRefs(),
    decodeNames: new Set(),
    pages: new Set(),
    needsJoinCsv: false,
    hasNested: false,
  };
  const classes: string[] = [];
  const paramInterfaces: string[] = [];

  const emit = (res: Resource, segments: string[]) => {
    if (res.resources?.length) uses.hasNested = true;
    classes.push(emitClass(res, segments, ctx, uses, paramInterfaces));
    for (const sub of res.resources || []) emit(sub, [...segments, sub.name]);
  };
  emit(resource, [resource.name]);

  const coreImports = ['RequestOptions', ...(uses.hasNested ? ['APIClient'] : [])].sort();
  const lines = [
    `import { APIResource } from '../core/resource';`,
    `import type { ${coreImports.join(', ')} } from '../core/request';`,
  ];
  if (uses.needsJoinCsv) lines.push(`import { joinCsv } from '../core/request';`);
  if (uses.pages.size) {
    const names = (['CursorPage', 'OffsetPage', 'Page'] as const).filter((n) => uses.pages.has(n));
    lines.push(`import { ${names.join(', ')} } from '../core/pagination';`);
  }
  const modelNames = uses.refs.sorted();
  if (modelNames.length) lines.push(`import type { ${modelNames.join(', ')} } from '../models';`);
  if (uses.decodeNames.size) lines.push(`import { ${[...uses.decodeNames].sort().join(', ')} } from '../models';`);

  const blocks = [lines.join('\n'), ...classes, ...paramInterfaces];
  return { path: `src/resources/${slug(resource.name) || 'resource'}.ts`, content: `${blocks.join('\n\n')}\n` };
}

function emitClass(
  resource: Resource,
  segments: string[],
  ctx: NodeCtx,
  uses: FileUses,
  paramInterfaces: string[],
): string {
  const cls = resourceClassName(segments);
  const subs = resource.resources || [];
  const members: string[] = [];

  // sub-resource fields + a constructor that wires them (openai-node's shape).
  if (subs.length) {
    for (const sub of subs) {
      members.push(`  readonly ${camelCase(sub.name)}: ${resourceClassName([...segments, sub.name])};`);
    }
    const ctorLines = ['    super(client);'];
    for (const sub of subs) {
      ctorLines.push(`    this.${camelCase(sub.name)} = new ${resourceClassName([...segments, sub.name])}(client);`);
    }
    members.push(`  constructor(client: APIClient) {\n${ctorLines.join('\n')}\n  }`);
  }

  for (const method of resource.methods || []) {
    members.push(emitMethod(segments, method, ctx, uses, paramInterfaces));
  }

  const doc = jsDoc(resource.description);
  const body = members.length ? `\n${members.join('\n\n')}\n` : '';
  return `${doc}export class ${cls} extends APIResource {${body}}`;
}

function emitMethod(
  segments: string[],
  method: Method,
  ctx: NodeCtx,
  uses: FileUses,
  paramInterfaces: string[],
): string {
  const op = planOperation(method, ctx.types);
  const name = nodeMethodName(method.action);
  const pathParams = op.paramGroups.path;
  const queryParams = op.paramGroups.query;
  const headerParams = op.paramGroups.header;
  const hasBody = op.hasBody;
  const rawContentType = op.binaryContentType;
  const bodyList = bodyFields(method.requestBody?.type, ctx.types);

  // The params object argument (body/query/header), openai-node's `PetCreateParams`.
  const hasParams = hasBody || queryParams.length > 0 || headerParams.length > 0;
  const argName = paramsArgName(hasBody, queryParams.length, headerParams.length);
  const paramsType = `${paramsQualifier(segments)}${pascalCase(method.action)}Params`;
  if (hasParams) {
    paramInterfaces.push(
      renderParamsInterface(paramsType, hasBody, bodyList, queryParams, headerParams, uses.refs),
    );
  }

  // signature: positional path args, then the params arg, then options.
  const args = pathParams.map((p) => `${camelCase(p.name)}: ${nodeType(p.type, uses.refs)}`);
  if (hasParams) {
    const optional = paramsRequired(hasBody, op.bodyRequired, queryParams, headerParams) ? '' : '?';
    args.push(`${argName}${optional}: ${paramsType}`);
  }
  args.push('options?: RequestOptions');

  // Whether the params argument is optional — expanded member accesses must then
  // use optional chaining (`query?.tags`) to stay strict-null safe.
  const argOptional = hasParams && !paramsRequired(hasBody, op.bodyRequired, queryParams, headerParams);

  // The request({ ... }) argument object (object-shorthand when the value is the arg itself).
  const entries = [`method: ${JSON.stringify(method.httpMethod.toUpperCase())}`, `path: ${pathExpr(method.path)}`];
  if (queryParams.length) {
    const qv = queryValue(argName, queryParams, uses, argOptional);
    entries.push(qv === 'query' ? 'query' : `query: ${qv}`);
  }
  const encoding = bodyEncoding(op, bodyList, ctx.types);
  if (hasBody) {
    const bv = bodyValue(argName, bodyList, queryParams.length + headerParams.length > 0, argOptional);
    entries.push(bv === 'body' ? 'body' : `body: ${bv}`);
  }
  const headerEntries = headerObjectEntries(rawContentType, argName, headerParams, argOptional);
  if (headerEntries) entries.push(`headers: ${headerEntries}`);
  if (hasBody && encoding !== 'json') entries.push(`encoding: ${JSON.stringify(encoding)}`);
  if (rawContentType) entries.push('raw: true');
  if (op.injectIdempotencyKey && ctx.behavior.idempotency.autoGenerateForPost) entries.push('idempotency: true');

  const argObject = `{ ${entries.join(', ')} }`;
  const { returnType, call } = returnPlan(op, argObject, uses, ctx);

  const guards = pathParamGuards(pathParams);
  const rawDoc = method.description ? jsDoc(method.description).trimEnd() : '';
  const docBlock = rawDoc ? `  ${rawDoc.replace(/\n/g, '\n  ')}\n` : '';
  return (
    `${docBlock}  ${name}(${args.join(', ')}): Promise<${returnType}> {\n` +
    guards +
    `${call}\n` +
    `  }`
  );
}

/**
 * The return type + the `return this._client.request(...)` statement. Binary
 * downloads return the raw Response; a paginated list wraps the envelope in its
 * vendored page container; a discriminator-mapped union routes through its
 * `decode<Union>` helper; everything else decodes to the primary response type.
 */
function returnPlan(
  op: OperationPlan,
  argObject: string,
  uses: FileUses,
  ctx: NodeCtx,
): { returnType: string; call: string } {
  const method = op.method;
  if (op.binaryContentType) {
    return { returnType: 'Response', call: `    return this._client.request<Response>(${argObject}, options);` };
  }
  if (op.pageName) {
    uses.pages.add(op.pageName);
    const item = nodeType(method.pagination?.itemType as TypeRef | undefined, uses.refs);
    return {
      returnType: `${op.pageName}<${item}>`,
      call:
        `    return this._client\n` +
        `      .request<unknown>(${argObject}, options)\n` +
        `      .then((raw) => ${op.pageName}.fromResponse<${item}>(raw));`,
    };
  }
  if (op.primaryResponse === 'none') {
    return { returnType: 'void', call: `    return this._client.request<void>(${argObject}, options);` };
  }
  const returnType = nodeType(method.primaryResponse, uses.refs);
  const decode = op.primaryResponse === 'union-mapped' ? unionDecodeFor(method.primaryResponse, ctx) : null;
  if (decode) {
    uses.decodeNames.add(decode);
    return {
      returnType,
      call:
        `    return this._client\n` +
        `      .request<unknown>(${argObject}, options)\n` +
        `      .then(${decode});`,
    };
  }
  return { returnType, call: `    return this._client.request<${returnType}>(${argObject}, options);` };
}

/** The `decode<Union>` helper name for a union-mapped primary-response ref, or null. */
function unionDecodeFor(ref: TypeRef | undefined, ctx: NodeCtx): string | null {
  if (ref?.kind !== 'ref' || !ref.name) return null;
  const named = ctx.types.get(ref.name);
  return named ? unionDecodeName(named) : null;
}

/** The params argument name, mirroring the Go emitter's planParams. */
function paramsArgName(hasBody: boolean, queryCount: number, headerCount: number): string {
  if (hasBody) return queryCount + headerCount > 0 ? 'params' : 'body';
  return queryCount > 0 ? 'query' : 'params';
}

function paramsRequired(hasBody: boolean, bodyRequired: boolean, query: Param[], header: Param[]): boolean {
  if (hasBody) return bodyRequired;
  return query.some((p) => p.required) || header.some((p) => p.required);
}

/** The body wire encoding: the IR encoding, upgraded to multipart for a json body carrying a binary field. */
function bodyEncoding(
  op: OperationPlan,
  bodyList: ReturnType<typeof bodyFields>,
  types: Map<string, NamedType>,
): 'json' | 'multipart' | 'form' {
  const encoding = op.encoding ?? 'json';
  if (encoding === 'json' && bodyList.some((f) => isBinaryTypeRef(f.type, types))) return 'multipart';
  return encoding;
}

function renderParamsInterface(
  typeName: string,
  hasBody: boolean,
  bodyList: ReturnType<typeof bodyFields>,
  queryParams: Param[],
  headerParams: Param[],
  refs: ModelRefs,
): string {
  const lines: string[] = [];
  if (hasBody) {
    for (const f of bodyList) {
      lines.push(`  ${propKey(f.name)}${f.required ? '' : '?'}: ${nodeType(f.type, refs)};`);
    }
  }
  // query/header params keyed by their LOGICAL name (mapped to the wire name at
  // request time), so a param like `tags` reads nicely even when its wire key is `tags[]`.
  for (const p of [...queryParams, ...headerParams]) {
    lines.push(`  ${propKey(p.name)}${p.required ? '' : '?'}: ${nodeType(p.type, refs)};`);
  }
  const body = lines.length ? `\n${lines.join('\n')}\n` : '';
  return `export interface ${typeName} {${body}}`;
}

/** The query value: the whole params object (query-only, no wire remapping), else a picked literal. */
function queryValue(argName: string, queryParams: Param[], uses: FileUses, optional: boolean): string {
  const needsExplicit = queryParams.some(
    (q) => (q.wireName && q.wireName !== q.name) || (q.type?.kind === 'array' && q.explode === false),
  );
  if (argName === 'query' && !needsExplicit) return 'query';
  const entries = queryParams.map((q) => {
    const wire = q.wireName ?? q.name;
    if (q.type?.kind === 'array' && q.explode === false) {
      uses.needsJoinCsv = true;
      return `${propKey(wire)}: joinCsv(${memberAccess(argName, q.name, optional)})`;
    }
    return `${propKey(wire)}: ${memberAccess(argName, q.name, optional)}`;
  });
  return `{ ${entries.join(', ')} }`;
}

/** The body value: the whole params object (body-only), else a picked literal (body fields). */
function bodyValue(
  argName: string,
  bodyList: ReturnType<typeof bodyFields>,
  mixed: boolean,
  optional: boolean,
): string {
  if (!mixed) return argName;
  const entries = bodyList.map((f) => `${propKey(f.name)}: ${memberAccess(argName, f.name, optional)}`);
  return `{ ${entries.join(', ')} }`;
}

/** The headers object literal (Accept for binary downloads + header params by wire name), or ''. */
function headerObjectEntries(
  rawContentType: string | null,
  argName: string,
  headerParams: Param[],
  optional: boolean,
): string {
  const entries: string[] = [];
  if (rawContentType) entries.push(`Accept: ${JSON.stringify(rawContentType)}`);
  for (const h of headerParams) {
    const wire = h.wireName ?? h.name;
    entries.push(`${propKey(wire)}: ${memberAccess(argName, h.name, optional)}`);
  }
  return entries.length ? `{ ${entries.join(', ')} }` : '';
}

/** `obj.key` (or `obj?.key` when the params arg is optional), quoting non-identifier keys. */
function memberAccess(obj: string, key: string, optional: boolean): string {
  const ident = /^[A-Za-z_$][A-Za-z0-9_$]*$/.test(key);
  if (optional) return ident ? `${obj}?.${key}` : `${obj}?.[${JSON.stringify(key)}]`;
  return ident ? `${obj}.${key}` : `${obj}[${JSON.stringify(key)}]`;
}

/** Reject an empty string for each required string path param before building the path. */
function pathParamGuards(pathParams: Param[]): string {
  let out = '';
  for (const p of pathParams) {
    if (p.type?.kind !== 'scalar' || p.type.scalar !== 'string' || p.required === false) continue;
    const name = camelCase(p.name);
    out += `    if (!${name}) {\n      throw new Error(${JSON.stringify(`missing required ${p.name} parameter`)});\n    }\n`;
  }
  return out;
}

/** The path as a string literal, or a template literal when it has params. */
function pathExpr(path: string): string {
  const p = path.startsWith('/') ? path : `/${path}`;
  if (!p.includes('{')) return JSON.stringify(p);
  const tmpl = p.replace(/\{([^}]+)\}/g, (_, w) => `\${${camelCase(w)}}`);
  return `\`${tmpl}\``;
}

File: crates/xyd_opensdk_node/src/resource.rs
//! `src/resources/<resource>.ts` + `src/resources/index.ts`. Ports `src/resource.ts`
//! (the generated-code slice; the Atlas type-reference view is out of fork scope).

use std::collections::BTreeSet;

use crate::ir::{Field, Method, NamedType, Param, Resource, TypeRef};
use crate::jsrt::{
    camel_case, js_doc, json_string, node_method_name, pascal_case, prop_key, singular_pascal_case,
    slug,
};
use crate::model::{body_fields, is_binary_type_ref, node_type, union_decode_name, ModelRefs};
use crate::plan::{plan_operation, OperationPlan, PrimaryResponseKind};

/// Per-file render context.
pub struct NodeCtx<'a> {
    pub types: Vec<&'a NamedType>,
    /// `behavior.idempotency.autoGenerateForPost` (default true).
    pub auto_generate_for_post: bool,
}

/// Per-file import trackers.
#[derive(Default)]
struct FileUses {
    refs: ModelRefs,
    decode_names: BTreeSet<String>,
    pages: BTreeSet<&'static str>,
    needs_join_csv: bool,
    has_nested: bool,
}

/// The globally-unique resource CLASS name, path-qualified by the full segment chain.
pub fn resource_class_name(segments: &[String]) -> String {
    segments.iter().map(|s| pascal_case(s)).collect()
}

fn params_qualifier(segments: &[String]) -> String {
    segments.iter().map(|s| singular_pascal_case(s)).collect()
}

/// The synthesized params-interface NAME (openai-node's `PetCreateParams` shape).
pub fn params_type_name(segments: &[String], action: &str) -> String {
    format!(
        "{}{}Params",
        params_qualifier(segments),
        pascal_case(action)
    )
}

/// Emit `src/resources/index.ts`: the barrel re-exporting every top-level resource file.
pub fn render_resources_index_file(resources: &[Resource]) -> String {
    let lines: Vec<String> = resources
        .iter()
        .map(|r| {
            let file = if slug(&r.name).is_empty() {
                "resource".to_string()
            } else {
                slug(&r.name)
            };
            format!("export * from './{file}';")
        })
        .collect();
    format!("{}\n", lines.join("\n"))
}

/// Emit one top-level resource (and its whole subtree) into a single file.
pub fn render_resource_file(resource: &Resource, ctx: &NodeCtx) -> (String, String) {
    let mut uses = FileUses::default();
    let mut classes: Vec<String> = Vec::new();
    let mut param_interfaces: Vec<String> = Vec::new();

    emit_tree(
        resource,
        std::slice::from_ref(&resource.name),
        ctx,
        &mut uses,
        &mut classes,
        &mut param_interfaces,
    );

    let mut core_imports = vec!["RequestOptions"];
    if uses.has_nested {
        core_imports.push("APIClient");
    }
    core_imports.sort_unstable();

    let mut lines = vec![
        "import { APIResource } from '../core/resource';".to_string(),
        format!(
            "import type {{ {} }} from '../core/request';",
            core_imports.join(", ")
        ),
    ];
    if uses.needs_join_csv {
        lines.push("import { joinCsv } from '../core/request';".to_string());
    }
    if !uses.pages.is_empty() {
        let names: Vec<&str> = ["CursorPage", "OffsetPage", "Page"]
            .into_iter()
            .filter(|n| uses.pages.contains(n))
            .collect();
        lines.push(format!(
            "import {{ {} }} from '../core/pagination';",
            names.join(", ")
        ));
    }
    let model_names: Vec<String> = uses.refs.iter().cloned().collect();
    if !model_names.is_empty() {
        lines.push(format!(
            "import type {{ {} }} from '../models';",
            model_names.join(", ")
        ));
    }
    if !uses.decode_names.is_empty() {
        let decodes: Vec<String> = uses.decode_names.iter().cloned().collect();
        lines.push(format!(
            "import {{ {} }} from '../models';",
            decodes.join(", ")
        ));
    }

    let mut blocks = vec![lines.join("\n")];
    blocks.extend(classes);
    blocks.extend(param_interfaces);
    let file = if slug(&resource.name).is_empty() {
        "resource".to_string()
    } else {
        slug(&resource.name)
    };
    (
        format!("src/resources/{file}.ts"),
        format!("{}\n", blocks.join("\n\n")),
    )
}

fn emit_tree(
    res: &Resource,
    segments: &[String],
    ctx: &NodeCtx,
    uses: &mut FileUses,
    classes: &mut Vec<String>,
    param_interfaces: &mut Vec<String>,
) {
    if !res.resources.is_empty() {
        uses.has_nested = true;
    }
    let cls = emit_class(res, segments, ctx, uses, param_interfaces);
    classes.push(cls);
    for sub in &res.resources {
        let mut seg = segments.to_vec();
        seg.push(sub.name.clone());
        emit_tree(sub, &seg, ctx, uses, classes, param_interfaces);
    }
}

fn emit_class(
    resource: &Resource,
    segments: &[String],
    ctx: &NodeCtx,
    uses: &mut FileUses,
    param_interfaces: &mut Vec<String>,
) -> String {
    let cls = resource_class_name(segments);
    let subs = &resource.resources;
    let mut members: Vec<String> = Vec::new();

    if !subs.is_empty() {
        for sub in subs {
            let mut seg = segments.to_vec();
            seg.push(sub.name.clone());
            members.push(format!(
                "  readonly {}: {};",
                camel_case(&sub.name),
                resource_class_name(&seg)
            ));
        }
        let mut ctor_lines = vec!["    super(client);".to_string()];
        for sub in subs {
            let mut seg = segments.to_vec();
            seg.push(sub.name.clone());
            ctor_lines.push(format!(
                "    this.{} = new {}(client);",
                camel_case(&sub.name),
                resource_class_name(&seg)
            ));
        }
        members.push(format!(
            "  constructor(client: APIClient) {{\n{}\n  }}",
            ctor_lines.join("\n")
        ));
    }

    for method in &resource.methods {
        members.push(emit_method(segments, method, ctx, uses, param_interfaces));
    }

    let doc = js_doc(resource.description.as_deref());
    let body = if members.is_empty() {
        String::new()
    } else {
        format!("\n{}\n", members.join("\n\n"))
    };
    format!("{doc}export class {cls} extends APIResource {{{body}}}")
}

fn emit_method(
    segments: &[String],
    method: &Method,
    ctx: &NodeCtx,
    uses: &mut FileUses,
    param_interfaces: &mut Vec<String>,
) -> String {
    let op = plan_operation(method, &ctx.types);
    let name = node_method_name(&method.action);
    let path_params = &method.path_params;
    let query_params = &method.query_params;
    let header_params = &method.header_params;
    let has_body = op.has_body;
    let raw_content_type = op.binary_content_type.clone();
    let body_list = body_fields(method.request_body.as_ref().map(|b| &b.ty), &ctx.types);

    let has_params = has_body || !query_params.is_empty() || !header_params.is_empty();
    let arg_name = params_arg_name(has_body, query_params.len(), header_params.len());
    let params_type = params_type_name(segments, &method.action);
    if has_params {
        param_interfaces.push(render_params_interface(
            &params_type,
            has_body,
            &body_list,
            query_params,
            header_params,
            &mut uses.refs,
        ));
    }

    let mut args: Vec<String> = path_params
        .iter()
        .map(|p| {
            format!(
                "{}: {}",
                camel_case(&p.name),
                node_type(Some(&p.ty), &mut uses.refs)
            )
        })
        .collect();
    if has_params {
        let optional = if params_required(has_body, op.body_required, query_params, header_params) {
            ""
        } else {
            "?"
        };
        args.push(format!("{arg_name}{optional}: {params_type}"));
    }
    args.push("options?: RequestOptions".to_string());

    let arg_optional =
        has_params && !params_required(has_body, op.body_required, query_params, header_params);

    let mut entries = vec![
        format!(
            "method: {}",
            json_string(&method.http_method.to_uppercase())
        ),
        format!("path: {}", path_expr(&method.path)),
    ];
    if !query_params.is_empty() {
        let qv = query_value(arg_name, query_params, uses, arg_optional);
        if qv == "query" {
            entries.push("query".to_string());
        } else {
            entries.push(format!("query: {qv}"));
        }
    }
    let encoding = body_encoding(&op, &body_list, &ctx.types);
    if has_body {
        let bv = body_value(
            arg_name,
            &body_list,
            !query_params.is_empty() || !header_params.is_empty(),
            arg_optional,
        );
        if bv == "body" {
            entries.push("body".to_string());
        } else {
            entries.push(format!("body: {bv}"));
        }
    }
    let header_entries = header_object_entries(
        raw_content_type.as_deref(),
        arg_name,
        header_params,
        arg_optional,
    );
    if !header_entries.is_empty() {
        entries.push(format!("headers: {header_entries}"));
    }
    if has_body && encoding != "json" {
        entries.push(format!("encoding: {}", json_string(encoding)));
    }
    if raw_content_type.is_some() {
        entries.push("raw: true".to_string());
    }
    if op.inject_idempotency_key && ctx.auto_generate_for_post {
        entries.push("idempotency: true".to_string());
    }

    let arg_object = format!("{{ {} }}", entries.join(", "));
    let (return_type, call) = return_plan(&op, method, &arg_object, uses, ctx);

    let guards = path_param_guards(path_params);
    let raw_doc = method
        .description
        .as_deref()
        .filter(|d| !d.is_empty())
        .map(|d| js_doc(Some(d)).trim_end().to_string())
        .unwrap_or_default();
    let doc_block = if raw_doc.is_empty() {
        String::new()
    } else {
        format!("  {}\n", raw_doc.replace('\n', "\n  "))
    };

    format!(
        "{doc_block}  {name}({args}): Promise<{return_type}> {{\n{guards}{call}\n  }}",
        args = args.join(", "),
    )
}

/// The return type + the `return this._client.request(...)` statement.
fn return_plan(
    op: &OperationPlan,
    method: &Method,
    arg_object: &str,
    uses: &mut FileUses,
    ctx: &NodeCtx,
) -> (String, String) {
    if op.binary_content_type.is_some() {
        return (
            "Response".to_string(),
            format!("    return this._client.request<Response>({arg_object}, options);"),
        );
    }
    if let Some(page_name) = op.page_name {
        uses.pages.insert(page_name);
        let item = node_type(
            method
                .pagination
                .as_ref()
                .and_then(|p| p.item_type.as_ref()),
            &mut uses.refs,
        );
        return (
            format!("{page_name}<{item}>"),
            format!(
                "    return this._client\n      .request<unknown>({arg_object}, options)\n      .then((raw) => {page_name}.fromResponse<{item}>(raw));"
            ),
        );
    }
    if op.primary_response == PrimaryResponseKind::None {
        return (
            "void".to_string(),
            format!("    return this._client.request<void>({arg_object}, options);"),
        );
    }
    let return_type = node_type(method.primary_response.as_ref(), &mut uses.refs);
    let decode = if op.primary_response == PrimaryResponseKind::UnionMapped {
        union_decode_for(method.primary_response.as_ref(), ctx)
    } else {
        None
    };
    if let Some(decode) = decode {
        uses.decode_names.insert(decode.clone());
        return (
            return_type,
            format!(
                "    return this._client\n      .request<unknown>({arg_object}, options)\n      .then({decode});"
            ),
        );
    }
    (
        return_type.clone(),
        format!("    return this._client.request<{return_type}>({arg_object}, options);"),
    )
}

fn union_decode_for(ref_: Option<&TypeRef>, ctx: &NodeCtx) -> Option<String> {
    let r = ref_?;
    if r.kind() != "ref" {
        return None;
    }
    let name = r.name.as_deref()?;
    let named = ctx.types.iter().find(|t| t.name == name)?;
    union_decode_name(named)
}

fn params_arg_name(has_body: bool, query_count: usize, header_count: usize) -> &'static str {
    if has_body {
        if query_count + header_count > 0 {
            "params"
        } else {
            "body"
        }
    } else if query_count > 0 {
        "query"
    } else {
        "params"
    }
}

fn params_required(has_body: bool, body_required: bool, query: &[Param], header: &[Param]) -> bool {
    if has_body {
        body_required
    } else {
        query.iter().any(|p| p.required_truthy()) || header.iter().any(|p| p.required_truthy())
    }
}

/// The body wire encoding: the IR encoding, upgraded to multipart for a json body
/// carrying a binary field.
fn body_encoding<'a>(op: &OperationPlan, body_list: &[&Field], types: &[&NamedType]) -> &'a str {
    let encoding = op.encoding.unwrap_or("json");
    if encoding == "json"
        && body_list
            .iter()
            .any(|f| is_binary_type_ref(Some(&f.ty), types, &mut BTreeSet::new()))
    {
        "multipart"
    } else {
        encoding
    }
}

fn render_params_interface(
    type_name: &str,
    has_body: bool,
    body_list: &[&Field],
    query_params: &[Param],
    header_params: &[Param],
    refs: &mut ModelRefs,
) -> String {
    let mut lines: Vec<String> = Vec::new();
    let mut seen: BTreeSet<String> = BTreeSet::new();
    // Body fields win over a same-keyed query/header param (dedupe by property key).
    if has_body {
        for f in body_list {
            push_param_field(
                &mut lines,
                &mut seen,
                &f.name,
                f.required == Some(true),
                &f.ty,
                refs,
            );
        }
    }
    for p in query_params.iter().chain(header_params.iter()) {
        push_param_field(
            &mut lines,
            &mut seen,
            &p.name,
            p.required_truthy(),
            &p.ty,
            refs,
        );
    }
    let body = if lines.is_empty() {
        String::new()
    } else {
        format!("\n{}\n", lines.join("\n"))
    };
    format!("export interface {type_name} {{{body}}}")
}

#[allow(clippy::too_many_arguments)]
fn push_param_field(
    lines: &mut Vec<String>,
    seen: &mut BTreeSet<String>,
    name: &str,
    required: bool,
    ty: &TypeRef,
    refs: &mut ModelRefs,
) {
    let key = prop_key(name);
    if seen.contains(&key) {
        return;
    }
    seen.insert(key.clone());
    lines.push(format!(
        "  {}{}: {};",
        key,
        if required { "" } else { "?" },
        node_type(Some(ty), refs)
    ));
}

fn query_value(
    arg_name: &str,
    query_params: &[Param],
    uses: &mut FileUses,
    optional: bool,
) -> String {
    let needs_explicit = query_params.iter().any(|q| {
        q.wire_name.as_deref().map(|w| w != q.name).unwrap_or(false)
            || (q.ty.kind() == "array" && q.explode == Some(false))
    });
    if arg_name == "query" && !needs_explicit {
        return "query".to_string();
    }
    let entries: Vec<String> = query_params
        .iter()
        .map(|q| {
            let wire = q.wire_name.as_deref().unwrap_or(&q.name);
            if q.ty.kind() == "array" && q.explode == Some(false) {
                uses.needs_join_csv = true;
                format!(
                    "{}: joinCsv({})",
                    prop_key(wire),
                    member_access(arg_name, &q.name, optional)
                )
            } else {
                format!(
                    "{}: {}",
                    prop_key(wire),
                    member_access(arg_name, &q.name, optional)
                )
            }
        })
        .collect();
    format!("{{ {} }}", entries.join(", "))
}

fn body_value(arg_name: &str, body_list: &[&Field], mixed: bool, optional: bool) -> String {
    if !mixed {
        return arg_name.to_string();
    }
    let entries: Vec<String> = body_list
        .iter()
        .map(|f| {
            format!(
                "{}: {}",
                prop_key(&f.name),
                member_access(arg_name, &f.name, optional)
            )
        })
        .collect();
    format!("{{ {} }}", entries.join(", "))
}

fn header_object_entries(
    raw_content_type: Option<&str>,
    arg_name: &str,
    header_params: &[Param],
    optional: bool,
) -> String {
    let mut entries: Vec<String> = Vec::new();
    if let Some(ct) = raw_content_type {
        entries.push(format!("Accept: {}", json_string(ct)));
    }
    for h in header_params {
        let wire = h.wire_name.as_deref().unwrap_or(&h.name);
        entries.push(format!(
            "{}: {}",
            prop_key(wire),
            member_access(arg_name, &h.name, optional)
        ));
    }
    if entries.is_empty() {
        String::new()
    } else {
        format!("{{ {} }}", entries.join(", "))
    }
}

fn is_js_ident(key: &str) -> bool {
    let mut chars = key.chars();
    match chars.next() {
        Some(c) if c.is_ascii_alphabetic() || c == '_' || c == '$' => {}
        _ => return false,
    }
    chars.all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '$')
}

fn member_access(obj: &str, key: &str, optional: bool) -> String {
    let ident = is_js_ident(key);
    if optional {
        if ident {
            format!("{obj}?.{key}")
        } else {
            format!("{obj}?.[{}]", json_string(key))
        }
    } else if ident {
        format!("{obj}.{key}")
    } else {
        format!("{obj}[{}]", json_string(key))
    }
}

fn path_param_guards(path_params: &[Param]) -> String {
    let mut out = String::new();
    for p in path_params {
        let is_string_scalar = p.ty.kind() == "scalar" && p.ty.scalar.as_deref() == Some("string");
        if !is_string_scalar || p.required == Some(false) {
            continue;
        }
        let name = camel_case(&p.name);
        out.push_str(&format!(
            "    if (!{name}) {{\n      throw new Error({});\n    }}\n",
            json_string(&format!("missing required {} parameter", p.name))
        ));
    }
    out
}

fn path_expr(path: &str) -> String {
    let p = if path.starts_with('/') {
        path.to_string()
    } else {
        format!("/{path}")
    };
    if !p.contains('{') {
        return json_string(&p);
    }
    // Replace `{name}` -> `${camelCase(name)}` inside a template literal.
    let mut tmpl = String::new();
    let mut rest = p.as_str();
    while let Some(start) = rest.find('{') {
        tmpl.push_str(&rest[..start]);
        let after = &rest[start + 1..];
        if let Some(end) = after.find('}') {
            let inner = &after[..end];
            tmpl.push_str(&format!("${{{}}}", camel_case(inner)));
            rest = &after[end + 1..];
        } else {
            tmpl.push_str(&rest[start..]);
            rest = "";
        }
    }
    tmpl.push_str(rest);
    format!("`{tmpl}`")
}
